ibc-apps/ics20-transfer/types/src/denom.rs

```diff
@@ -386,18 +401,86 @@ mod tests {
+        PrefixedDenom::from_str("/uatom").expect("no error");
+        PrefixedDenom::from_str("//uatom").expect("no error");
+        PrefixedDenom::from_str("transfer/").expect("no error");
+        PrefixedDenom::from_str("transfer/atom").expect("no error");
+        PrefixedDenom::from_str("(transfer)/channel-0/uatom").expect("no error");
+        PrefixedDenom::from_str("transfer/(channel-0)/uatom").expect("no error");
+
+        Ok(())
+    }
+
+    #[rstest]
+    #[case(
+        "transfer/channel-75",
+        "factory/stars16da2uus9zrsy83h23ur42v3lglg5rmyrpqnju4/dust"
+    )]
+    #[case(
+        "transfer/channel-75/transfer/channel-123/transfer/channel-1023/transfer/channel-0",
+        "factory/stars16da2uus9zrsy83h23ur42v3lglg5rmyrpqnju4/dust"
+    )]
+    #[case(
+        "transfer/channel-75/transfer/channel-123/transfer/channel-1023/transfer/channel-0",
+        "//////////////////////dust"
+    )]
+    // https://github.com/cosmos/ibc-go/blob/e2ad31975f2ede592912b86346b5ebf055c9e05f/modules/apps/transfer/types/trace_test.go#L17-L38
+    #[case("", "uatom")]
+    #[case("", "uatom/")]
+    #[case("", "gamm/pool/1")]
+    #[case("", "gamm//pool//1")]
+    #[case("transfer/channel-1", "uatom")]
+    #[case("customtransfer/channel-1", "uatom")]
+    #[case("transfer/channel-1", "uatom/")]
+    #[case(
+        "transfer/channel-1",
+        "erc20/0x85bcBCd7e79Ec36f4fBBDc54F90C643d921151AA"
+    )]
+    #[case("transfer/channel-1", "gamm/pool/1")]
+    #[case("transfer/channel-1", "gamm//pool//1")]
+    #[case("transfer/channel-1/transfer/channel-2", "uatom")]
+    #[case("customtransfer/channel-1/alternativetransfer/channel-2", "uatom")]
+    #[case("", "transfer/uatom")]
+    #[case("", "transfer//uatom")]
+    #[case("", "channel-1/transfer/uatom")]
+    #[case("", "uatom/transfer")]
+    #[case("", "transfer/channel-1")]
+    #[case("transfer/channel-1", "transfer")]
+    #[case("", "transfer/channelToA/uatom")]
+    fn test_strange_but_accepted_prefixed_denom(
+        #[case] prefix: &str,
+        #[case] denom: &str,
+    ) -> Result<(), TokenTransferError> {
+        let pd_s = if prefix.is_empty() {
+            denom.to_owned()
+        } else {
+            format!("{prefix}/{denom}")
+        };
+        let pd = PrefixedDenom::from_str(&pd_s)?;
+
+        assert_eq!(pd.trace_path.to_string(), prefix);
+        assert_eq!(pd.base_denom.to_string(), denom);
 
         Ok(())
     }
 
+    #[rstest]
+    #[case("")]
+    #[case("transfer/channel-1")]
+    #[case("transfer/channel-1/transfer/channel-2")]
+    #[should_panic(expected = "EmptyBaseDenom")]
+    fn test_prefixed_empty_base_denom(#[case] prefix: &str) {
+        let pd_s = if prefix.is_empty() {
+            "".to_owned()
+        } else {
+            format!("{prefix}/")
+        };
+        PrefixedDenom::from_str(&pd_s).expect("error");
+    }
+
     #[test]
     fn test_denom_trace() -> Result<(), TokenTransferError> {
```

Comment:
As far as I see, this `test_denom_trace` is covered by the added test cases. Should this be removed?

Reply:
02d57c8 b1804b0